src/wallet/wallet.c

```diff
@@ -92,128 +96,152 @@ static void get_address_path(uint32_t account, bool change, uint32_t index, char
   }
 }
 
-#if 0  // TODO, remove or refactor?
-static transaction_payload_t* wallet_build_transaction(iota_wallet_t* w, bool change, uint32_t sender_index,
-                                                       byte_t receiver[], uint64_t balance, char const index[],
-                                                       byte_t data[], size_t data_len) {
-  char tmp_addr[IOTA_ADDRESS_HEX_BYTES + 1] = {};
-  char addr_path[IOTA_ACCOUNT_PATH_MAX] = {};
-  byte_t send_addr[ED25519_ADDRESS_BYTES] = {};
-  byte_t tmp_tx_id[TRANSACTION_ID_BYTES] = {};
-  ed25519_keypair_t addr_keypair = {};
-  res_outputs_id_t* outputs_res = NULL;
-  transaction_payload_t* tx_payload = NULL;
-  int ret = -1;
-
-  // TODO loop over start and end addresses
-  // get address keypair and address
-  get_address_path(w->account_index, change, sender_index, addr_path, sizeof(addr_path));
-
-  if (address_keypair_from_path(w->seed, sizeof(w->seed), addr_path, &addr_keypair) != 0) {
-    printf("[%s:%d] Cannot get address keypair\n", __func__, __LINE__);
-    goto done;
-  }
-
-  if (address_from_ed25519_pub(addr_keypair.pub, send_addr) != 0) {
-    printf("[%s:%d] Cannot get sending address \n", __func__, __LINE__);
-    goto done;
-  }
-
-  // get outputs
-  bin_2_hex(send_addr, sizeof(send_addr), tmp_addr, sizeof(tmp_addr));
-  if (!(outputs_res = res_outputs_new())) {
-    printf("[%s:%d] Err: invalid length of path\n", __func__, __LINE__);
+// create basic unspent outputs
+static utxo_outputs_list_t* basic_outputs_from_address(iota_wallet_t* w, transaction_essence_t* essence,
+                                                       ed25519_keypair_t* sender_key, address_t* send_addr,
+                                                       uint64_t send_amount, signing_data_list_t** sign_data,
+                                                       uint64_t* output_amount) {
+  int ret = 0;
+  char bech32_addr[BIN_TO_HEX_STR_BYTES(ADDRESS_MAX_BYTES)] = {};
+  // query output IDs from Indexer by bech32 address
+  res_outputs_id_t* res_id = res_outputs_new();
+  if (res_id == NULL) {
+    printf("[%s:%d] allocate outputs response failed\n", __func__, __LINE__);
     return NULL;
   }
 
-  // FIXME : get_outputs_id method now accepts query params list
-  if (get_outputs_id(&w->endpoint, false, tmp_addr, outputs_res) != 0) {
-    printf("[%s:%d] Err: get outputs from address failed\n", __func__, __LINE__);
-    goto done;
-  }
-
-  if (outputs_res->is_error) {
-    printf("[%s:%d] Error get outputs from addr: %s\n", __func__, __LINE__, outputs_res->u.error->msg);
-    goto done;
-  }
+  if (address_to_bech32(send_addr, w->bech32HRP, bech32_addr, sizeof(bech32_addr)) == 0) {
+    outputs_query_list_t* query_param = outputs_query_list_new();
+    ret = outputs_query_list_add(&query_param, QUERY_PARAM_ADDRESS, bech32_addr);
+    if (ret != 0) {
+      printf("[%s:%d] add query params failed\n", __func__, __LINE__);
+      outputs_query_list_free(query_param);
```

Comment:
`res_id` should also be freed.

```suggestion
      outputs_query_list_free(query_param);
      res_outputs_free(res_id);
```